File: tarefa-1/pi_serie.c
/* ============================================================================
 * pi_serie.c - Aproximacao de PI por series matematicas
 *
 * Objetivo: variar o numero de iteracoes, medir o tempo de execucao e observar
 * como o erro em relacao ao valor real de PI diminui conforme gastamos mais
 * processamento.
 *
 * Duas series sao comparadas:
 *   Leibniz-Gregory : pi = 4 * (1 - 1/3 + 1/5 - 1/7 + ...)      erro ~ 1/N
 *   Nilakantha      : pi = 3 + 4/(2*3*4) - 4/(4*5*6) + ...      erro ~ 1/(4N^3)
 *
 * Compilar:  gcc -O2 -Wall -Wextra -std=c99 pi_serie.c -o pi_serie -lm
 * Executar:  ./pi_serie          (vai ate 10^9 iteracoes)
 *            ./pi_serie 7        (vai ate 10^7 iteracoes)
 * ==========================================================================*/

#define _POSIX_C_SOURCE 199309L   /* libera clock_gettime() no Linux/macOS */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#if defined(_WIN32)
#include <windows.h>
#endif

/* Valor de referencia de PI com mais digitos do que um double consegue guardar
 * (um double tem ~15-17 digitos decimais significativos). Nao usamos M_PI
 * porque ele nao faz parte do C padrao - so existe como extensao. */
static const double PI_REF = 3.14159265358979323846;

/* Quantos valores de N vamos testar, no maximo (10^1 ate 10^9). */
#define MAX_EXPOENTE 9

/* ---------------------------------------------------------------------------
 * MEDICAO DE TEMPO
 *
 * Existem DOIS tempos diferentes, e a diferenca entre eles e' o coracao da
 * disciplina de programacao paralela:
 *
 *   tempo de parede (wall time) : o relogio da parede, o que o usuario espera
 *   tempo de CPU               : soma do tempo que os nucleos ficaram ocupados
 *
 * Em codigo sequencial os dois sao praticamente iguais. Em codigo paralelo com
 * 8 threads, 1 segundo de parede pode custar 8 segundos de CPU. O speedup e'
 * sempre calculado com tempo de PAREDE.
 * -------------------------------------------------------------------------*/

/* Tempo de parede, em segundos, com resolucao de nanosegundos.
 * O #if e' so portabilidade: Windows e POSIX tem APIs diferentes. */
static double tempo_parede(void)
{
#if defined(_WIN32)
    LARGE_INTEGER freq, agora;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&agora);
    return (double)agora.QuadPart / (double)freq.QuadPart;
#else
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1e-9;
#endif
}

/* Tempo de CPU consumido pelo processo, em segundos.
 * clock() devolve "ticks"; CLOCKS_PER_SEC converte para segundos. */
static double tempo_cpu(void)
{
    return (double)clock() / (double)CLOCKS_PER_SEC;
}

/* ---------------------------------------------------------------------------
 * AS SERIES
 * -------------------------------------------------------------------------*/

/* Serie de Leibniz-Gregory:
 *
 *     pi/4 = 1 - 1/3 + 1/5 - 1/7 + 1/9 - ...
 *
 * O termo k (comecando em 0) e' (-1)^k / (2k+1).
 *
 * Note que NAO usamos pow(-1.0, k) para alternar o sinal: pow() e' uma chamada
 * de funcao cara (dezenas de ciclos) e seria executada bilhoes de vezes.
 * Multiplicar uma variavel por -1 custa 1 ciclo. Trocar pow() por isso deixa o
 * laco varias vezes mais rapido sem mudar o resultado. */
static double serie_leibniz(long long n)
{
    double soma  = 0.0;
    double sinal = 1.0;
    long long k;

    for (k = 0; k < n; k++) {
        soma += sinal / (double)(2 * k + 1);
        sinal = -sinal;
    }
    return 4.0 * soma;
}

/* Serie de Nilakantha:
 *
 *     pi = 3 + 4/(2*3*4) - 4/(4*5*6) + 4/(6*7*8) - ...
 *
 * O termo k (comecando em 1) usa a = 2k e vale (-1)^(k+1) * 4/(a(a+1)(a+2)).
 * O denominador cresce com k^3, entao o erro cai com 1/N^3 - muito mais rapido
 * que Leibniz. Serve para mostrar que trocar de ALGORITMO costuma valer mais
 * que empilhar iteracoes (ou nucleos) em cima de um algoritmo ruim. */
static double serie_nilakantha(long long n)
{
    double soma  = 3.0;
    double sinal = 1.0;
    long long k;

    for (k = 1; k <= n; k++) {
        double a = 2.0 * (double)k;
        soma += sinal * 4.0 / (a * (a + 1.0) * (a + 2.0));
        sinal = -sinal;
    }
    return soma;
}

/* ---------------------------------------------------------------------------
 * APOIO
 * -------------------------------------------------------------------------*/

/* Quantas casas decimais estao corretas, aproximadamente.
 * Se o erro e' 1e-5, entao -log10(1e-5) = 5 casas corretas. */
static double casas_corretas(double erro)
{
    if (erro <= 0.0) return 17.0;          /* erro zero: limite do double */
    return -log10(erro);
}

/* Executa uma serie para um N, mede os tempos e imprime a linha da tabela.
 * Recebe um ponteiro para funcao para nao duplicar todo esse codigo de medicao
 * uma vez para Leibniz e outra para Nilakantha. */
static void executa(const char *nome, double (*serie)(long long), long long n,
                    double *erro_saida, double *tempo_saida)
{
    double t0_parede, t1_parede, t0_cpu, t1_cpu;
    double aprox, erro;

    t0_parede = tempo_parede();
    t0_cpu    = tempo_cpu();

    aprox = serie(n);                       /* <-- o trabalho de verdade */

    t1_cpu    = tempo_cpu();
    t1_parede = tempo_parede();

    erro = fabs(PI_REF - aprox);

    printf("  %-11s %12lld  %.15f  %10.3e  %6.2f  %9.4f  %9.4f\n",
           nome, n, aprox, erro, casas_corretas(erro),
           t1_parede - t0_parede, t1_cpu - t0_cpu);

    *erro_saida  = erro;
    *tempo_saida = t1_parede - t0_parede;
}

/* ---------------------------------------------------------------------------
 * PROGRAMA PRINCIPAL
 * -------------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
    int expoente_max = MAX_EXPOENTE;
    int e, i;

    /* Arrays para guardar os resultados e analisar a convergencia no final. */
    long long ns[MAX_EXPOENTE + 1];
    double erro_leib[MAX_EXPOENTE + 1], tempo_leib[MAX_EXPOENTE + 1];
    double erro_nila[MAX_EXPOENTE + 1], tempo_nila[MAX_EXPOENTE + 1];
    int qtd = 0;

    /* argv[1] opcional: expoente maximo, para nao esperar muito em teste. */
    if (argc > 1) {
        expoente_max = atoi(argv[1]);
        if (expoente_max < 1 || expoente_max > MAX_EXPOENTE) {
            fprintf(stderr, "Uso: %s [expoente entre 1 e %d]\n",
                    argv[0], MAX_EXPOENTE);
            return 1;
        }
    }

    printf("Valor de referencia de PI : %.15f\n", PI_REF);
    printf("Precisao do tipo double   : ~15-17 digitos decimais\n\n");

    printf("  %-11s %12s  %-17s %10s  %6s  %9s  %9s\n",
           "SERIE", "N", "APROXIMACAO", "ERRO ABS", "CASAS", "T.PAREDE", "T.CPU");
    printf("  ---------------------------------------------------------"
           "-------------------------------------\n");

    for (e = 1; e <= expoente_max; e++) {
        long long n = 1;
        int j;
        for (j = 0; j < e; j++) n *= 10;    /* n = 10^e sem usar pow() */

        ns[qtd] = n;
        executa("Leibniz",    serie_leibniz,    n, &erro_leib[qtd], &tempo_leib[qtd]);
        executa("Nilakantha", serie_nilakantha, n, &erro_nila[qtd], &tempo_nila[qtd]);
        printf("\n");
        qtd++;
    }

    /* ---- Analise da convergencia -----------------------------------------
     * A cada passo multiplicamos N por 10. Se o erro cai por um fator ~10, a
     * convergencia e' de 1a ordem (erro ~ 1/N). Se cai por ~1000, e' de 3a
     * ordem (erro ~ 1/N^3). O tempo, em ambos os casos, cresce ~10x, porque o
     * custo e' O(N) nas duas series. */
    printf("Fator de reducao do erro e de aumento do tempo a cada 10x em N:\n\n");
    printf("  %14s  %12s  %12s  %12s\n",
           "N", "ERRO/10x LEIB", "ERRO/10x NILA", "TEMPO/10x");
    printf("  ------------------------------------------------------------\n");
    for (i = 1; i < qtd; i++) {
        double fator_leib  = erro_leib[i - 1]  / erro_leib[i];
        double fator_nila  = erro_nila[i - 1]  / erro_nila[i];
        double fator_tempo = (tempo_leib[i - 1] > 0.0)
                             ? tempo_leib[i] / tempo_leib[i - 1] : 0.0;
        printf("  %14lld  %12.1f  %12.1f  %12.1f\n",
               ns[i], fator_leib, fator_nila, fator_tempo);
    }

    printf("\nLeitura: com Leibniz, 10x mais trabalho compra ~1 casa decimal.\n");
    printf("Com Nilakantha, o mesmo trabalho compra ~3 casas - ate a serie\n");
    printf("bater no limite de precisao do double (~1e-16), onde mais\n");
    printf("iteracoes so acumulam erro de arredondamento e nao melhoram nada.\n");

    return 0;
}

File: tarefa-1/pi_serie.ts
/* ============================================================================
 * pi_serie.ts - Aproximacao de PI por series matematicas (porte do pi_serie.c)
 *
 * Mesma estrutura do programa em C, de proposito: mesmas series, mesmos N,
 * mesma medicao de tempo. So assim a comparacao entre as duas linguagens e'
 * honesta - se a estrutura mudasse, estariamos medindo a mudanca, nao a
 * linguagem.
 *
 * Executar:  node --experimental-strip-types pi_serie.ts
 *            node --experimental-strip-types pi_serie.ts 7   (vai ate 10^7)
 *
 * O Node 22 executa TypeScript direto apagando os tipos (type stripping).
 * Nao ha compilacao, nao ha tsc, nao ha bundler - os tipos sao apagados e o
 * JavaScript resultante e' entregue ao V8.
 * ==========================================================================*/

/* Valor de referencia de PI. Diferente do C, aqui Math.PI E' padrao da
 * linguagem - e vale exatamente o mesmo double que a constante do C. */
const PI_REF: number = Math.PI;

const MAX_EXPOENTE = 9;

/* ---------------------------------------------------------------------------
 * MEDICAO DE TEMPO
 *
 * Mesma distincao do C, com as APIs do Node:
 *   performance.now()    -> tempo de parede, em milissegundos, monotonico
 *   process.cpuUsage()   -> tempo de CPU (user + system), em microssegundos
 *
 * performance.now() e' o equivalente direto do QueryPerformanceCounter /
 * clock_gettime(CLOCK_MONOTONIC). NAO usamos Date.now(): ele e' o relogio de
 * parede do sistema, tem resolucao de ~1ms e pode ANDAR PARA TRAS quando o NTP
 * ajusta o horario da maquina - o que produziria tempos negativos.
 * -------------------------------------------------------------------------*/

function tempoParede(): number {
  return performance.now() / 1000; // ms -> s
}

function tempoCpu(): number {
  const u = process.cpuUsage();
  return (u.user + u.system) / 1_000_000; // microssegundos -> s
}

/* ---------------------------------------------------------------------------
 * AS SERIES
 *
 * Ponto crucial para a comparacao: o `number` do JavaScript E' um double IEEE
 * 754 de 64 bits, exatamente o mesmo tipo do `double` do C. Portanto os valores
 * numericos aqui saem IDENTICOS aos do C, bit a bit. O que muda entre as duas
 * linguagens e' so o TEMPO, nunca a acuracia.
 *
 * Note que os contadores sao `number`, nao `bigint`. Seria tentador usar bigint
 * para "inteiros de verdade", mas bigint aloca objetos no heap e e' ~50x mais
 * lento. Um number representa inteiros exatos ate 2^53 (~9e15), muito acima dos
 * 10^9 que precisamos.
 * -------------------------------------------------------------------------*/

function serieLeibniz(n: number): number {
  let soma = 0.0;
  let sinal = 1.0;

  for (let k = 0; k < n; k++) {
    soma += sinal / (2 * k + 1);
    sinal = -sinal;
  }
  return 4.0 * soma;
}

function serieNilakantha(n: number): number {
  let soma = 3.0;
  let sinal = 1.0;

  for (let k = 1; k <= n; k++) {
    const a = 2.0 * k;
    soma += (sinal * 4.0) / (a * (a + 1.0) * (a + 2.0));
    sinal = -sinal;
  }
  return soma;
}

/* ---------------------------------------------------------------------------
 * APOIO
 * -------------------------------------------------------------------------*/

function casasCorretas(erro: number): number {
  if (erro <= 0) return 17;
  return -Math.log10(erro);
}

/* Alinhamento de colunas. Em C o printf faz isso com "%12lld"; em JS/TS nao
 * existe printf, entao usamos padStart/padEnd na mao. */
const dir = (s: string | number, w: number) => String(s).padStart(w);
const esq = (s: string | number, w: number) => String(s).padEnd(w);

interface Resultado {
  erro: number;
  tempo: number;
}

function executa(
  nome: string,
  serie: (n: number) => number,
  n: number,
): Resultado {
  const t0Parede = tempoParede();
  const t0Cpu = tempoCpu();

  const aprox = serie(n); // <-- o trabalho de verdade

  const t1Cpu = tempoCpu();
  const t1Parede = tempoParede();

  const erro = Math.abs(PI_REF - aprox);
  const tempo = t1Parede - t0Parede;

  console.log(
    `  ${esq(nome, 11)} ${dir(n, 12)}  ${aprox.toFixed(15)}  ` +
      `${dir(erro.toExponential(3), 10)}  ${dir(casasCorretas(erro).toFixed(2), 6)}  ` +
      `${dir(tempo.toFixed(4), 9)}  ${dir((t1Cpu - t0Cpu).toFixed(4), 9)}`,
  );

  return { erro, tempo };
}

/* ---------------------------------------------------------------------------
 * AQUECIMENTO (warm-up) - nao existe equivalente em C
 *
 * O V8 e' um compilador JIT: na primeira vez, a funcao roda interpretada
 * (Ignition). Depois de algumas milhares de iteracoes ele detecta que o codigo
 * e' "quente", infere os tipos observados e recompila para codigo de maquina
 * otimizado (Maglev/TurboFan).
 *
 * Sem este aquecimento, as primeiras linhas da tabela mediriam o interpretador
 * e as ultimas o codigo otimizado - e a tabela mostraria uma aceleracao que e'
 * artefato do JIT, nao propriedade do algoritmo. Em C isso nao existe: o codigo
 * ja nasce compilado e a primeira iteracao roda tao rapido quanto a bilionesima.
 * -------------------------------------------------------------------------*/
function aquece(): void {
  for (let i = 0; i < 5; i++) {
    serieLeibniz(200_000);
    serieNilakantha(200_000);
  }
}

/* ---------------------------------------------------------------------------
 * PROGRAMA PRINCIPAL
 * -------------------------------------------------------------------------*/
function main(): void {
  // process.argv[0] = node, [1] = script, [2] = primeiro argumento do usuario
  const arg = process.argv[2];
  let expoenteMax = MAX_EXPOENTE;

  if (arg !== undefined) {
    expoenteMax = Number.parseInt(arg, 10);
    if (!Number.isInteger(expoenteMax) || expoenteMax < 1 || expoenteMax > MAX_EXPOENTE) {
      console.error(`Uso: node --experimental-strip-types pi_serie.ts [1..${MAX_EXPOENTE}]`);
      process.exit(1);
    }
  }

  console.log(`Runtime                   : Node ${process.version} (V8 ${process.versions.v8})`);
  console.log(`Valor de referencia de PI : ${PI_REF.toFixed(15)}`);
  console.log(`Precisao do tipo number   : double IEEE-754, ~15-17 digitos\n`);

  aquece();

  console.log(
    `  ${esq('SERIE', 11)} ${dir('N', 12)}  ${esq('APROXIMACAO', 17)} ` +
      `${dir('ERRO ABS', 10)}  ${dir('CASAS', 6)}  ${dir('T.PAREDE', 9)}  ${dir('T.CPU', 9)}`,
  );
  console.log('  ' + '-'.repeat(92));

  const ns: number[] = [];
  const leib: Resultado[] = [];
  const nila: Resultado[] = [];

  for (let e = 1; e <= expoenteMax; e++) {
    const n = 10 ** e;

    ns.push(n);
    leib.push(executa('Leibniz', serieLeibniz, n));
    nila.push(executa('Nilakantha', serieNilakantha, n));
    console.log('');
  }

  console.log('Fator de reducao do erro e de aumento do tempo a cada 10x em N:\n');
  console.log(
    `  ${dir('N', 14)}  ${dir('ERRO/10x LEIB', 13)}  ${dir('ERRO/10x NILA', 13)}  ${dir('TEMPO/10x', 12)}`,
  );
  console.log('  ' + '-'.repeat(60));

  for (let i = 1; i < ns.length; i++) {
    const fLeib = leib[i - 1].erro / leib[i].erro;
    const fNila = nila[i - 1].erro / nila[i].erro;
    const fTempo = leib[i - 1].tempo > 0 ? leib[i].tempo / leib[i - 1].tempo : 0;

    console.log(
      `  ${dir(ns[i], 14)}  ${dir(fLeib.toFixed(1), 13)}  ${dir(fNila.toFixed(1), 13)}  ${dir(fTempo.toFixed(1), 12)}`,
    );
  }

  console.log('\nLeitura: com Leibniz, 10x mais trabalho compra ~1 casa decimal.');
  console.log('Com Nilakantha, o mesmo trabalho compra ~3 casas - ate a serie');
  console.log('bater no limite de precisao do double (~1e-16), onde mais');
  console.log('iteracoes so acumulam erro de arredondamento e nao melhoram nada.');
}

main();
